src/digraph.rs

```diff
@@ -2265,6 +2266,150 @@ impl PyDiGraph {
         Ok(NodeMap { node_map: out_map })
     }
 
+    /// Substitute a set of nodes with a single node.
+    ///
+    /// :param set to_replace: A set of nodes to be removed and replaced
+    ///     by the specified node.
+    /// :param int node: An existing node to use as the replacement.
+    /// :param bool force_check_cycle: If set to ``True``, validates
+    ///     that the substitution will not introduce cycles before
+    ///     modifying the graph, even if cycle check is disabled for
+    ///     this instance of :class:`retworkx.PyDiGraph`. Otherwise,
+    ///     the value of member ``check_cycle`` is inherited.
+    /// :raises DAGWouldCycle: The cycle check is enabled and the
+    ///     substitution would introduce cycle(s).
+    /// :raises ValueError: ``to_replace`` is empty or ``node``
+    ///     has existing connections.
+    #[pyo3(
+        text_signature = "(self, to_replace, node, /, force_check_cycle=None)"
+    )]
+    fn substitute_nodes_with_node(
+        &mut self,
+        py: Python,
+        to_replace: HashSet<usize>,
+        node: usize,
+        force_check_cycle: Option<bool>,
+    ) -> PyResult<()> {
+        fn can_contract(
+            graph: &StablePyGraph<Directed>,
+            nodes: &HashSet<NodeIndex>,
+        ) -> bool {
+            let mut visited = graph.visit_map();
+            let mut visit_next = VecDeque::new();
+
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            for node in nodes {
+                for edge in graph.edges(*node) {
+                    if !nodes.contains(&edge.target()) {
+                        visit_next.push_back(edge.target());
+                    }
+                }
+            }
+
+            // DFS. If we encounter any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            while let Some(decendant) = visit_next.pop_back() {
+                if visited.is_visited(&decendant) {
+                    continue;
+                }
+
+                for edge in graph.edges(decendant) {
+                    if nodes.contains(&edge.target()) {
+                        // we found a path back to `nodes`
+                        return false;
+                    }
+
+                    visit_next.push_back(edge.target());
+                }
+
+                visited.visit(decendant);
+            }
```

Comment:
This works fine here, as an alternative we could use [`petgraph::visit::Dfs`](https://docs.rs/petgraph/0.6.0/petgraph/visit/struct.Dfs.html)
```suggestion
            let mut dfs = Dfs::from_parts(visit_next, visited);
            while let Some(node) = dfs.next(graph) {
                if nodes.contains(&node) {
                    // we found a path back to `nodes`
                    return false;
                }
            }
```

Reply:
Wow, very cool!